packages/core/src/libraries/tenant-usage.ts

```diff
@@ -0,0 +1,111 @@
+import { ConnectorType } from '@logto/schemas';
+import { sql } from '@silverhand/slonik';
+import type { CommonQueryMethods } from '@silverhand/slonik';
+
+import type { ConnectorLibrary } from '#src/libraries/connector.js';
+
+import TenantUsageQuery from '../queries/tenant-usage/index.js';
+import {
+  tenantUsageGuard,
+  selfComputedSubscriptionUsageGuard,
+  type SelfComputedTenantUsage,
+} from '../queries/tenant-usage/types.js';
+
+export class SelfComputedUsage {
+  private readonly sqlBuilder: TenantUsageQuery;
+
+  constructor(
+    public readonly pool: CommonQueryMethods,
+    public readonly connectorLibrary: ConnectorLibrary,
+    public readonly tenantId: string
+  ) {
+    this.sqlBuilder = new TenantUsageQuery(tenantId);
+  }
+
+  public async getTenantUsage(): Promise<{ usage: SelfComputedTenantUsage }> {
+    const rawUsage = await this.getRawTenantUsage();
+
+    const connectors = await this.connectorLibrary.getLogtoConnectors();
+    const socialConnectors = connectors.filter(
+      (connector) => connector.type === ConnectorType.Social
+    );
+
+    const unparsedUsage: SelfComputedTenantUsage = {
+      applicationsLimit: Number(rawUsage.applicationsLimit),
```

Comment:
nit: Can add these string -> number logic directly at the zod guard level using transform.